### ISSUE TITLE
How to name parameter dimensions?

### ISSUE THREAD

**Issue by Hvass-Labs** [scikit-optimize/scikit-optimize#570]
Hello, I applaud your effort with this library and I'm sorry to bother you with such a simple question.

I would like to use skopt with Keras and TensorFlow for optimization of hyper-parameters of a Convolutional Neural Network.

I have installed skopt v.0.4 using pip.

Ideally I would like to name the hyper-parameters only once when I define the search-space, and when skopt calls my "fitness" function, it gives named arguments so I don't have to unpack a list of parameters, which is error-prone.

It appears that `BayesSearchCV` can take a dict of named parameter-dimensions. But I don't think I can use `BayesSearchCV` because I need to run `fit_generator()` from Keras instead of `fit()`, because I want to use larger datasets from the harddisk that don't fit in memory.

So I think I have to use `gp_minimize()` but this does not take a dict of named parameters, instead it takes a list of dimensions, which can be defined in various ways.

I tried doing the following to name the parameters:

    from skopt.space import Real, Categorical, Integer

    dimensions = [Real(name='learning_rate', low=1e-6, high=1e-3, prior='log-uniform'),
	          Integer(name='num_nodes', low=10, high=256),
	          Categorical(name='activation', categories=['relu', 'sigmoid'])]

And then I would call:

    from skopt import gp_minimize

    res = gp_minimize(func=f, dimensions=dimensions)

The docs say that `Integer`, `Real` and `Categorical` can take a `name` argument:

https://scikit-optimize.github.io/space/space.m.html#skopt.space.space.Integer

But I get the following error:

    TypeError: __init__() got an unexpected keyword argument 'name'

Going into the source-code for `space.py` we see that `name` is indeed missing:

    class Real(Dimension):
        def __init__(self, low, high, prior="uniform", transform=None):

Is this because I have installed an older version of skopt using pip?

Could you advise how to construct named dimensions in the search-space. Ideally I would only have to name each dimension once when I define it, and then in my fitness function  I would like to get named arguments somehow, so I don't need to unpack a list.

Thanks!

**Comment by betatim** [comment]
You need to install scikit-optimize from GitHub directly for that feature. It did not yet exist when 0.4 was released.

However I do not think that it will call your objective function with named arguments. I like the idea, so if someone wanted to contribute support for that it would be welcome. One thing that comes to mind straight away is that it would have to continue to work for functions that have no named arguments for backwards compatibility. Not quite sure how much python magic we would have to perform to pull that off.

**Comment by Hvass-Labs** [comment]
Thanks for the quick reply!

Yes it would be very nice if my function would get called as `f(learning_rate=X, num_nodes=Y, activation=Z)` or it could be a dict that would be easy to unpack:

    def f(params):
        learning_rate = params['learning_rate']
        num_nodes = params['num_nodes']
        activation = params['activation']

This is not quite as elegant, but certainly better than unpacking from a list.

In my own SwarmOps library on Meta-Optimization which is a closely related idea to hyper-parameter optimization, I also unpacked from a list, but it is hard to understand the code and it is error-prone. :-)

I'm completely new to skopt so I don't think I can implement this in skopt. But perhaps you can switch behaviour of skopt depending on how the dimensions are defined: If it is a dict with named dimensions, then it will call the function with named args, otherwise it will call it as usual for backwards compatibility. Or you can make a breaking change - I think this one would be justified because it would be a big improvement to the API.

If someone implements this feature then please drop a note in this thread if you remember, so I get a notification.

I don't know if you want to close this thread or keep it open?

**Comment by betatim** [comment]
Let's keep this open to discuss this and we will reference it in the PR that adds this (or close it).

I'd be -0 on breaking the API but +1 on finding out how much magic is needed to detect if the objective supports named arguments or not. Then basing the calling convention on that.

**Comment by betatim** [comment]
The inspect module gives us (robust) access to the argument names and values: https://docs.python.org/3/library/inspect.html#inspect.signature. Unfortunately in python2 the support isn't quite as robust yet.

**Comment by Hvass-Labs** [comment]
I've got it working with Keras and TensorFlow and so far I think skopt is awesome! I am very impressed with it!

I have a few comments / questions. I know that you are just a couple of people developing this so I don't expect this to get fixed overnight :-)

(1) I'm now trying to plot the results using e.g. `plot_objective()`. The docs say it can take named dimensions, but that is not in the actual code for v. 0.4, so I'm guessing this is a v. 0.5 feature as well?

Would it be possible for you to update the pip-package to 0.5? I am currently trying to piece together a TensorFlow tutorial on hyper-parameter optimization and it needs to be easy to install the packages I use.

(2) It appears the axes on the plots are named X_0, X_1, etc. If the dimensions were named, then the plots should use these names so they are easier to read.

(3) I cannot use `plot_evaluations()` it gives an error: `TypeError: '>=' not supported between instances of 'numpy.ndarray' and 'float'` which appears to come from the call to `ax[i, i].hist()` inside `plot_evaluations()`. Perhaps it is crashing because I have a categorical dimension? If I could name the dimensions to plot, then perhaps I could omit the categorical dimension from `plot_evaluations()` so it works?

(4) Regarding named arguments to the objective functions that we discussed above. I understand that you are generally not too happy about making breaking changes to the API. But there are a few things you may consider: (A) Your current users know that skopt is in development and it may have breaking changes in the future. Your current users are "early adopters" so they are probably fine with that. (B) If you continue improving skopt then the vast majority of your users are still to come in the future. You have probably only reached max 5-10% of the people that will eventually use skopt. (C) New people are more likely to use skopt if the API is even simpler / better. (D) It is much better for everyone to have a clean and simple API that improves and has breaking changes occasionally, than an API that tries to be backwards compatible at all costs.

The author of Keras recently wrote a blog-post about API design and I found it to be very inspirational. Perhaps you may like it as well:

https://blog.keras.io/author/francois-chollet.html

I would like to finish my tutorial in just a few days. I suppose I could delay it, but if it's going to be months before these issues get resolved, then I'll probably hack my way around these issues for now. Then perhaps I would also have to fix my tutorial later on if you decide to make breaking changes to the API :-)

I hope I'm not being too pushy :-) As I have worked in the related field of Meta-Optimization 10 years ago, I am really excited to see that skopt is so much better than what I made back then! With a few tweaks to the API, skopt would be an amazing tool!

**Comment by Hvass-Labs** [comment]
I should have mentioned that I also get an error in `plot_objective()`. It plots a few of the contour-plots and then the last row of plots is blank and I get this error:

    TypeError: ufunc 'multiply' did not contain a loop with signature matching types dtype('<U32') dtype('<U32') dtype('<U32')

The problem appears to be in the function `partial_dependence()` with this line:

    yi = np.linspace(bounds[0], bounds[1], n_points)

My guess is that it is the categorical dimension that is causing this problem again?

It really would be great if I could put these plots in my tutorial. Would this get fixed if I could pass named dimensions to `plot_objective()` using the v. 0.5 code, so I could omit the categorical dimension?

**Comment by Hvass-Labs** [comment]
It would also be great if there was a built-in function in skopt for plotting the Gaussian process as done here:

https://scikit-optimize.github.io/notebooks/bayesian-optimization.html

**Comment by betatim** [comment]
Meta comment: could you make separate issues for each of your points (except (4) which is this issue)? I think they are all valid so we should track them in their own issue, otherwise they get lost. That would be great.

**Comment by betatim** [comment]
On releasing v0.5 and `pip`: would be good to hear the opinion of the other maintainers but I think we'd like to fix a few of the issues on the roadmap for the next release before having v0.5. However progress is slow because everyone is busy with other things (as usual). Thoughts on making a quick release with what we have so far?

In the mean time you can install directly from `master` in a way that is reproducible with:
```
$ pip install https://github.com/scikit-optimize/scikit-optimize/archive/1677188ee6de3365f7925de780dd2f31e5e34ee4.zip
```
replacing `1677188ee6de3365f7925de780dd2f31e5e34ee4` with a recent commit hash.

**Comment by Hvass-Labs** [comment]
My tutorial on TensorFlow / Keras / skopt is now done except for the plots from skopt. I think it would improve the tutorial substantially if those plots were also in, so I am willing to wait for those issues to get fixed.

I think I need to give the plot-functions the names of the dimensions to plot, so it doesn't crash on the categorical dimension I have. Is that right? It may not be necessary to open other issues if v. 0.5 takes care of all this.

I don't need the other feature with named-arguments right now, but the plots would be nice.

I will have to travel in a few days and it might be a while before I can work on the tutorial again. But I don't suppose you can update the pip-package to v.0.5 so I can make the plots already today or tomorrow. So I'll probably have to wait and finish the tutorial later. It's a pity because I think there would be very large interest in this tutorial, which would also give you an even larger user-base.

I think it would be too complicated to ask my audience to install from latest github source. The pip-package would be best. But why are you waiting so long to update it anyway? Wouldn't it be better to update the pip-package whenever you have something substantial? If you have good unit-tests for everything then it should be fine to update it regularly instead of waiting months for some bug-fix.

I thought a little more about the named parameters being passed to the objective function: If possible I think the API should continue to support the parameters packed as a list and be backwards compatible as you said. The reason is that there might be applications where this makes most sense - e.g. the benchmark problems you have in your demos obviously don't have named parameters. But it would be so much better for hyper-parameter optimization of neural networks if the parameters were named, so I don't have to unpack the list. I have made a boiler-plate function for that, but it is easy to make a mistake when you have many parameters.

I really like this library you have made, it is awesome - kudos! :-)

**Comment by Hvass-Labs** [comment]
I forgot to close this. It is solved in PR #592.
